### PR TITLE
[RFC]  Reference attribute linking to product variants

### PR DESCRIPTION
**Pull request by IKarbowiak**
The goal is to allow staff users to link the reference attribute to the product variant. 

## General assumptions
- Staff users with `MANAGE_PRODUCTS`  permission will be able to assign the product variant as a reference attribute.
- The product variant could be assigned only to attributes with `productVariant` set as the `entity type` .

**The current reference attribute assumptions that stay unchanged:**
- The `entityType` of reference attribute must be set during the attribute creation.
- The `enityType`  of reference attribute cannot be changed.
- To allow setting the reference attribute, the reference attribute must be first assigned to the `ProductType` .
- Setting the reference attribute value is the same as for other attributes. The field responsible for that is `references`  in the `AttributeValueInput` .

## Database changes
- Extend `AttributeValue`  with not required `reference_variant: ForeignKey`  field
- Extend `AttributeEntityType`  with `PRODUCT_VARIANT`

## API changes
Extend the `AttributeEntityTypeEnum` with `PRODUCT_VARIANT`

```graphql
enum AttributeEntityTypeEnum {
  PAGE
  PRODUCT
  PRODUCT_VARIANT
}
```

# Impact

* [ ] New migrations
* [ ] New/Updated API fields or mutations
* [ ] Deprecated API fields or mutations
* [ ] Removed API types, fields, or mutations
* [ ] Documentation needs to be updated

# Pull Request Checklist

* [ ] Privileged queries and mutations are guarded by proper permission checks
* [ ] Database queries are optimized and the number of queries is constant
* [ ] Database migration files are up to date
* [ ] The changes are tested
* [ ] GraphQL schema and type definitions are up to date
* [ ] Changes are mentioned in the changelog

### CODE REVIEW

**Comment by patrys** [comment]
Task linked: [SALEOR-7922 Reference attribute linking to product variants](https://app.clickup.com/t/2549495/SALEOR-7922)

File: saleor/graphql/schema.graphql [hunk 1]
```diff
@@ -15799,7 +15800,7 @@ input ProductCreateInput {
 
 input AttributeValueInput {
   """ID of the selected attribute."""
-  id: ID
+  id: ID!
```
Comment (IKarbowiak) [inline saleor/graphql/schema.graphql:15803]: I made it required as a requirement of this field was tested in utils anyway,

File: saleor/graphql/product/tests/test_product.py [hunk 2]
```diff
@@ -8684,32 +8825,6 @@ def test_update_product_with_non_existing_attribute(
     ]
 
 
-def test_update_product_with_no_attribute_slug_or_id(
-    staff_api_client, product, permission_manage_products, color_attribute
-):
```
Comment (IKarbowiak) [inline saleor/graphql/product/tests/test_product.py:8689]: Right now the `id` requirement is ensured by graphql.
There was no possibility to provide a `slug` for a longer time.

**Review by fowczarek — APPROVED** [review]

File: saleor/graphql/attribute/utils.py [hunk 3]
```diff
@@ -52,6 +52,8 @@ class AttrValuesInput:
 T_INPUT_MAP = List[Tuple[attribute_models.Attribute, AttrValuesInput]]
 T_ERROR_DICT = Dict[Tuple[str, str], List[str]]
 
+EntityTypeData = namedtuple("EntityTypeData", ["model", "entity_field", "value_field"])
```
Comment (maarcingebala) [inline saleor/graphql/attribute/utils.py:55]: Minor improvement - could we rename `entity_field` to something like `name_field`? If I understand correctly, it represents a field that for each type should be used to get the reference's name, right?

File: saleor/graphql/schema.graphql [hunk 4]
```diff
@@ -4828,6 +4828,7 @@ enum AttributeInputTypeEnum {
 enum AttributeEntityTypeEnum {
   PAGE
   PRODUCT
+  PRODUCTVARIANT
```
Comment (maarcingebala) [inline saleor/graphql/schema.graphql:4831]: `PRODUCT_VARIANT`

File: saleor/graphql/attribute/utils.py (hunk 3 above)
Comment (IKarbowiak) [reply saleor/graphql/attribute/utils.py:55]: Yep. I'll change it

**Review by maarcingebala — APPROVED** [review]

**Review by maarcingebala — APPROVED** [review]

**Review by fowczarek — APPROVED** [review]

**Pull request merged**